Rwjs/cppmutex

robsimmonds opened the pull request:
Removed remaining TBB components.

robsimmonds committed:
Code needed to replace TBB group contexts.

robsimmonds committed:
Changes to make the backend use CPP threads rather than TBB threads for event processing.

robsimmonds committed:
Put back code removed while testing.

robsimmonds committed:
Replaced last TBB group context.

robsimmonds committed:
Replaced last tbb group context.

robsimmonds committed:
Remove commment.

robsimmonds committed:
Fixed reference count issue.

robsimmonds committed:
Merge branch 'dev' into rwjs/cppthreads

robsimmonds committed:
Made OnMessageTask destructor virtual.

robsimmonds committed:
Removed remaining TBB components.

robsimmonds committed:
Components needed to replace TBB classes.

robsimmonds committed:
Merge branch 'dev' into rwjs/cppmutex

robsimmonds committed:
Fixed merge conflict.

kswang1029 commented:
@acdo2002 @ajm-asiaa could you confirm and check the failures of the ICD tests about tile browser? 

Update: initial investigation suggests this should be nothing to do with this branch.

kswang1029 commented:
my initial test result is positive. I don't see major issues. I will try few more times in the next two days. 

Just one time I saw a crash but I cannot reproduce it:

> libc++abi.dylib: terminating with uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument
libc++abi.dylib: terminating with uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument
Abort trap: 6

crash log from OS


```
Process:               carta_backend [43395]
Path:                  /Users/USER/*/carta_backend
Identifier:            carta_backend
Version:               0
Code Type:             X86-64 (Native)
Parent Process:        bash [957]
Responsible:           Terminal [617]
User ID:               501

Date/Time:             2021-11-01 13:37:08.281 +0800
OS Version:            Mac OS X 10.15.7 (19H1519)
Report Version:        12
Bridge OS Version:     6.0 (19P548)
Anonymous UUID:        E10256D2-ED0E-DD37-2159-E177CC612F78

Sleep/Wake UUID:       E1BF5853-A4F4-434D-AA66-CBCAC502F3DC

Time Awake Since Boot: 93000 seconds
Time Since Wake:       1500 seconds

System Integrity Protection: enabled

Crashed Thread:        3

Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Application Specific Information:
abort() called
terminating with uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument

Thread 0:: Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	0x00007fff6bb8ed2e _kernelrpc_mach_vm_deallocate_trap + 10
1   libsystem_kernel.dylib        	0x00007fff6bb90399 mach_vm_deallocate + 64
2   libsystem_malloc.dylib        	0x00007fff6bc11866 mvm_deallocate_pages + 70
3   libsystem_malloc.dylib        	0x00007fff6bc0ff46 free_large + 877
4   carta_backend                 	0x000000010f92d7db Frame::~Frame() + 843 (Frame.h:87)
5   carta_backend                 	0x000000010f92d457 std::__1::__shared_ptr_pointer<Frame*, std::__1::default_delete<Frame>, std::__1::allocator<Frame> >::__on_zero_shared() + 23 (memory:3541)
6   carta_backend                 	0x000000010f9184ae Session::DeleteFrame(int) + 190 (Session.cc:546)
7   carta_backend                 	0x000000010fb0f32b carta::SessionManager::OnMessage(uWS::WebSocket<false, true, PerSocketData>*, std::__1::basic_string_view<char, std::__1::char_traits<char> >, uWS::OpCode) + 2443
8   carta_backend                 	0x000000010fb1e904 uWS::WebSocketContext<false, true, PerSocketData>::handleFragment(char*, unsigned long, unsigned int, int, bool, uWS::WebSocketState<true>*, void*) + 1300 (WebSocketContext.h:93)
9   carta_backend                 	0x000000010fb1db97 bool uWS::WebSocketProtocol<true, uWS::WebSocketContext<false, true, PerSocketData> >::consumeMessage<6u, unsigned char>(unsigned char, char*&, unsigned int&, uWS::WebSocketState<true>*, void*) + 711 (WebSocketProtocol.h:328)
10  carta_backend                 	0x000000010fb1d820 uWS::WebSocketProtocol<true, uWS::WebSocketContext<false, true, PerSocketData> >::consume(char*, unsigned int, uWS::WebSocketState<true>*, void*) + 272
11  carta_backend                 	0x000000010fb1d6be auto uWS::WebSocketContext<false, true, PerSocketData>::init()::'lambda'(auto*, char*, int)::operator()<us_socket_t>(auto*, char*, int) const + 142 (WebSocketContext.h:289)
12  carta_backend                 	0x000000010fc659b5 us_loop_run + 133 (epoll_kqueue.c:129)
13  carta_backend                 	0x000000010fb11287 carta::SessionManager::RunApp() + 343
14  carta_backend                 	0x000000010f8dfb7b main + 4923 (Main.cc:215)
15  libdyld.dylib                 	0x00007fff6ba4dcc9 start + 1

Thread 1:
0   libsystem_kernel.dylib        	0x00007fff6bb91882 __psynch_cvwait + 10
1   libsystem_pthread.dylib       	0x00007fff6bc52425 _pthread_cond_wait + 698
2   libc++.1.dylib                	0x00007fff68d1e623 std::__1::condition_variable::__do_timed_wait(std::__1::unique_lock<std::__1::mutex>&, std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l> > >) + 93
3   carta_backend                 	0x000000010f94c61b std::__1::cv_status std::__1::condition_variable::wait_until<std::__1::chrono::steady_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l> > >(std::__1::unique_lock<std::__1::mutex>&, std::__1::chrono::time_point<std::__1::chrono::steady_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l> > > const&) + 171 (__mutex_base:426)
4   carta_backend                 	0x000000010f94c4f4 spdlog::details::periodic_worker::periodic_worker(std::__1::function<void ()> const&, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1l> >)::'lambda'()::operator()() const + 100 (periodic_worker-inl.h:25)
5   carta_backend                 	0x000000010f94c41d void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, spdlog::details::periodic_worker::periodic_worker(std::__1::function<void ()> const&, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1l> >)::'lambda'()> >(void*) + 45 (thread:286)
6   libsystem_pthread.dylib       	0x00007fff6bc52109 _pthread_start + 148
7   libsystem_pthread.dylib       	0x00007fff6bc4db8b thread_start + 15

Thread 2:
0   libsystem_kernel.dylib        	0x00007fff6bb9533a __pthread_kill + 10
1   libsystem_pthread.dylib       	0x00007fff6bc51e60 pthread_kill + 430
2   libsystem_c.dylib             	0x00007fff6bb1c808 abort + 120
3   libc++abi.dylib               	0x00007fff68d79458 abort_message + 231
4   libc++abi.dylib               	0x00007fff68d6a8a7 demangling_terminate_handler() + 238
5   libobjc.A.dylib               	0x00007fff6a8a75b1 _objc_terminate() + 104
6   libc++abi.dylib               	0x00007fff68d78887 std::__terminate(void (*)()) + 8
7   libc++abi.dylib               	0x00007fff68d7b1a2 __cxxabiv1::failed_throw(__cxxabiv1::__cxa_exception*) + 27
8   libc++abi.dylib               	0x00007fff68d7b169 __cxa_throw + 113
9   libc++.1.dylib                	0x00007fff68d5555b std::__1::__throw_system_error(int, char const*) + 77
10  libc++.1.dylib                	0x00007fff68d4c54d std::__1::mutex::lock() + 29
11  carta_backend                 	0x000000010f943cd9 queuing_rw_mutex::writer_leave() + 25 (Concurrency.h:88)
12  carta_backend                 	0x000000010f93b64c Frame::FillSpatialProfileData(PointXy, std::__1::vector<CARTA::SetSpatialRequirements_SpatialConfig, std::__1::allocator<CARTA::SetSpatialRequirements_SpatialConfig> >, std::__1::vector<CARTA::SpatialProfileData, std::__1::allocator<CARTA::SpatialProfileData> >&) + 5820 (Frame.cc:1251)
13  carta_backend                 	0x000000010f939f24 Frame::FillSpatialProfileData(std::__1::vector<CARTA::SpatialProfileData, std::__1::allocator<CARTA::SpatialProfileData> >&) + 68
14  carta_backend                 	0x000000010f91bf5c Session::SendSpatialProfileData(int, int) + 1244 (Session.cc:1439)
15  carta_backend                 	0x000000010f91b9e6 Session::OnSetCursor(CARTA::SetCursor const&, unsigned int) + 1014 (Session.cc:678)
16  carta_backend                 	0x000000010fb24fde FileSettings::ExecuteOne(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int) + 382
17  carta_backend                 	0x000000010fb236e8 SetCursorTask::execute() + 56 (OnMessageTask.cc:29)
18  carta_backend                 	0x000000010fb33c65 void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, carta::ThreadManager::StartEventHandlingThreads(int)::$_0> >(void*) + 181 (thread:284)
19  libsystem_pthread.dylib       	0x00007fff6bc52109 _pthread_start + 148
20  libsystem_pthread.dylib       	0x00007fff6bc4db8b thread_start + 15

Thread 3 Crashed:
0   libsystem_kernel.dylib        	0x00007fff6bb9533a __pthread_kill + 10
1   libsystem_pthread.dylib       	0x00007fff6bc51e60 pthread_kill + 430
2   libsystem_c.dylib             	0x00007fff6bb1c808 abort + 120
3   libc++abi.dylib               	0x00007fff68d79458 abort_message + 231
4   libc++abi.dylib               	0x00007fff68d6a8a7 demangling_terminate_handler() + 238
5   libobjc.A.dylib               	0x00007fff6a8a75b1 _objc_terminate() + 104
6   libc++abi.dylib               	0x00007fff68d78887 std::__terminate(void (*)()) + 8
7   libc++abi.dylib               	0x00007fff68d7b1a2 __cxxabiv1::failed_throw(__cxxabiv1::__cxa_exception*) + 27
8   libc++abi.dylib               	0x00007fff68d7b169 __cxa_throw + 113
9   libc++.1.dylib                	0x00007fff68d5555b std::__1::__throw_system_error(int, char const*) + 77
10  libc++.1.dylib                	0x00007fff68d4c54d std::__1::mutex::lock() + 29
11  carta_backend                 	0x000000010f943cd9 queuing_rw_mutex::writer_leave() + 25 (Concurrency.h:88)
12  carta_backend                 	0x000000010f93b64c Frame::FillSpatialProfileData(PointXy, std::__1::vector<CARTA::SetSpatialRequirements_SpatialConfig, std::__1::allocator<CARTA::SetSpatialRequirements_SpatialConfig> >, std::__1::vector<CARTA::SpatialProfileData, std::__1::allocator<CARTA::SpatialProfileData> >&) + 5820 (Frame.cc:1251)
13  carta_backend                 	0x000000010f939f24 Frame::FillSpatialProfileData(std::__1::vector<CARTA::SpatialProfileData, std::__1::allocator<CARTA::SpatialProfileData> >&) + 68
14  carta_backend                 	0x000000010f91bf5c Session::SendSpatialProfileData(int, int) + 1244 (Session.cc:1439)
15  carta_backend                 	0x000000010f91eabf Session::OnSetSpatialRequirements(CARTA::SetSpatialRequirements const&) + 1055 (Session.cc:853)
16  carta_backend                 	0x000000010fb196b4 GeneralMessageTask<CARTA::SetSpatialRequirements>::execute() + 20 (OnMessageTask.tcc:36)
17  carta_backend                 	0x000000010fb33c65 void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, carta::ThreadManager::StartEventHandlingThreads(int)::$_0> >(void*) + 181 (thread:284)
18  libsystem_pthread.dylib       	0x00007fff6bc52109 _pthread_start + 148
19  libsystem_pthread.dylib       	0x00007fff6bc4db8b thread_start + 15

Thread 3 crashed with X86 Thread State (64-bit):
  rax: 0x0000000000000000  rbx: 0x000070000cd66000  rcx: 0x000070000cd65228  rdx: 0x0000000000000000
  rdi: 0x0000000000002803  rsi: 0x0000000000000006  rbp: 0x000070000cd65250  rsp: 0x000070000cd65228
   r8: 0x000070000cd650f0   r9: 0x000070000cd652c0  r10: 0x000070000cd66000  r11: 0x0000000000000246
  r12: 0x0000000000002803  r13: 0x0000003000000008  r14: 0x0000000000000006  r15: 0x0000000000000016
  rip: 0x00007fff6bb9533a  rfl: 0x0000000000000246  cr2: 0x00000001105b4008
  
Logical CPU:     0
Error Code:      0x0100000c
Trap Number:     133
```

kswang1029 commented:
Here is another crash but I cannot reproduce it. This is when I have a frozen cursor and unfreeze-then-move it to see spatial profiles. 

```
Thread 2 Crashed:
0   carta_backend                 	0x000000010b270d53 std::__1::__hash_table<std::__1::__hash_value_type<int, std::__1::mutex>, std::__1::__unordered_map_hasher<int, std::__1::__hash_value_type<int, std::__1::mutex>, std::__1::hash<int>, true>, std::__1::__unordered_map_equal<int, std::__1::__hash_value_type<int, std::__1::mutex>, std::__1::equal_to<int>, true>, std::__1::allocator<std::__1::__hash_value_type<int, std::__1::mutex> > >::remove(std::__1::__hash_const_iterator<std::__1::__hash_node<std::__1::__hash_value_type<int, std::__1::mutex>, void*>*>) + 83 (__hash_table:2671)
1   carta_backend                 	0x000000010b467f97 FileSettings::ExecuteOne(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int) + 311 (FileSettings.cc:36)
2   carta_backend                 	0x000000010b4666e8 SetCursorTask::execute() + 56 (OnMessageTask.cc:29)
3   carta_backend                 	0x000000010b476c65 void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, carta::ThreadManager::StartEventHandlingThreads(int)::$_0> >(void*) + 181 (thread:284)
4   libsystem_pthread.dylib       	0x00007fff6bc52109 _pthread_start + 148
5   libsystem_pthread.dylib       	0x00007fff6bc4db8b thread_start + 15
```

robsimmonds committed:
Fixed typo / bug...

robsimmonds committed:
Merge branch 'dev' into rwjs/cppmutex

acdo2002 commented:
I test the performance in this branch, so far I didn't see any significant change in carta performance.
the comparison to dev branch can be checked in [here](https://carta.asiaa.sinica.edu.tw/performance/)

robsimmonds committed:
Merge branch 'dev' into rwjs/cppmutex

confluence commented on src/Concurrency.h:
```diff
@@ -0,0 +1,144 @@
+#include <mutex>
+#include <shared_mutex>
+
+#define __USE_TBB__ 0
```
Should this be removed? It's unused.

confluence commented on src/Concurrency.h:
```diff
@@ -0,0 +1,144 @@
+        if (_writer_count > 0) {
+            _readers_cv.wait(lock);
+        } else {
+            ++_reader_count;
```
Should this be incremented conditionally here? It's decremented unconditionally in `reader_leave`.

confluence commented on src/Concurrency.h:
```diff
@@ -0,0 +1,144 @@
+            _mtx_list.push_front(wait_mtx);
+            wait_mtx->lock();
+        }
+    }
```
`_writer_count` is never incremented.

confluence commented on src/Concurrency.h:
```diff
@@ -0,0 +1,144 @@
+                _rwmtx->reader_leave();
+            }
+            _active = false;
+        }
```
This could call `release` to avoid repetition. This would be consistent with `~concurrent_queue` calling `clear`.

confluence commented on src/SessionManager/SessionManager.cc:
```diff
@@ -52,12 +52,13 @@ void SessionManager::OnUpgrade(
         http_response->close();
         return;
     }
+    */
```
This should be uncommented.

confluence reviewed (changes requested):
I'm not sure if I'm following the logic of `queuing_rw_mutex`, but it looks like the private reader and writer counts aren't being incremented and decremented symmetrically. Other than that there's a small bit of avoidable code repetition, and the token check has been left disabled.

robsimmonds committed:
Put back token check.

robsimmonds committed:
Added some missing logic.

robsimmonds committed:
Removed unused definition.

robsimmonds committed:
Refactored to remove some repeated code.

robsimmonds committed:
Fixed another logic issue.

robsimmonds committed:
Added condition veriable for writers and removed list that was no longer needed.

robsimmonds committed:
Small optimisation.

confluence reviewed (changes requested):
There's a `shared_mutex` in STL since C++17 which we could perhaps use here instead of the custom mutex classes.

If we keep the custom classes, we should rename them, since they no longer involve a queue.

robsimmonds commented:
> There's a `shared_mutex` in STL since C++17 which we could perhaps use here instead of the custom mutex classes.
> 
> If we keep the custom classes, we should rename them, since they no longer involve a queue.

Yes, if the STL ones will do what we need we should use them. I was trying to decide if we do need to order writes anywhere. If we do then the advantage of keeping our own implementation is we can add the queue.

pford commented:
> 
> 
> > There's a `shared_mutex` in STL since C++17 which we could perhaps use here instead of the custom mutex classes.
> > If we keep the custom classes, we should rename them, since they no longer involve a queue.
> 
> Yes, if the STL ones will do what we need we should use them. I was trying to decide if we do need to order writes anywhere. If we do then the advantage of keeping our own implementation is we can add the queue.

As the user mouses around the image, it is important to keep the **last** cursor setting.  I looked up the TBB queuing mutex, and it is fair: "lets threads through in the order they arrive" (https://docs.oneapi.io/versions/latest/onetbb/tbb_userguide/Mutex_Flavors.html).  In the [shared_mutex documentation](https://en.cppreference.com/w/cpp/named_req/SharedMutex) I read, it does not indicate whether it is fair or not.  This [stackoverflow question](https://stackoverflow.com/questions/17527621/is-stdmutex-fair) about std::mutex mentions the TBB mutex as a fair alternative to the likely unfair native mutex, and another answer suggests using conditional variables to enforce fairness.

robsimmonds committed:
Added back the queue!

robsimmonds committed:
Removed comment.

confluence reviewed (changes requested):
I keep forgetting -- please update the changelog.

pford commented:
Why is the Concurrency namespace a separate issue #957 rather than being fixed here?  Seems like a good opportunity to me.

pford reviewed (changes requested):
It would be helpful to have comments in Concurrency.h explaining the purpose of each class for developers using this code in the future.

robsimmonds committed:
Merge branch 'dev' into rwjs/cppmutex

robsimmonds committed:
Added carta namespace.
Chaged queuing_rw_mutex_local to queuing_rw_mutex_scoped which describes its purpose better.

robsimmonds committed:
Added the namespace to the mutexes.

robsimmonds committed:
Added namespace to the mutexes.

robsimmonds committed:
Removed mention of TBB.

robsimmonds committed:
Removed TBB lib.

robsimmonds committed:
Added that we no longer need TBB with this branch.

robsimmonds committed:
Merge branch 'dev' into rwjs/cppmutex

robsimmonds committed:
Added descriptions for the classes.

robsimmonds committed:
Moved some entries from "Added" to "Changed" section.

robsimmonds committed:
Fixed formatting.

robsimmonds committed:
Fixed formatting.

veggiesaurus commented on src/Session.h:
```diff
@@ -222,7 +221,7 @@ class Session {
 
     FileSettings _file_settings;
-    std::unordered_map<int, tbb::concurrent_queue<std::pair<CARTA::SetImageChannels, uint32_t>>> _set_channel_queues;
+    std::unordered_map<int, concurrent_queue<std::pair<CARTA::SetImageChannels, uint32_t>>> _set_channel_queues;
```
shouldn't this be `carta::concurrent_queue`? Since `Session` is not part of the `carta` namespace yet

veggiesaurus commented on src/Session.h:
```diff
@@ -318,9 +317,9 @@ class Session {
 
     // message queue <msg, compress>
-    tbb::concurrent_queue<std::pair<std::vector<char>, bool>> _out_msgs;
+    concurrent_queue<std::pair<std::vector<char>, bool>> _out_msgs;
```
same here

veggiesaurus reviewed (changes requested):
two minor issues I spotted

robsimmonds committed:
Fixed typos in comments.

robsimmonds committed:
Added namespace tags.

robsimmonds committed:
Added hyphen in thread-safe.

confluence reviewed (approved):
I think this is fine to merge now; we just need it to be tested again because the implementation has changed.

kswang1029 commented:
> I think this is fine to merge now; we just need it to be tested again because the implementation has changed.

allowing me some time to test it

confluence commented:
Adding a note that this now fixes #957 so that it can be closed automatically (probably).

kswang1029 commented:
Got one crash while playing around region spectral profiler by deleting regions:
libc++abi.dylib: terminating with uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument
Abort trap: 6

Thread 2 Crashed:
0   libsystem_kernel.dylib        	0x00007fff6bb9533a __pthread_kill + 10
1   libsystem_pthread.dylib       	0x00007fff6bc51e60 pthread_kill + 430
2   libsystem_c.dylib             	0x00007fff6bb1c808 abort + 120
3   libc++abi.dylib               	0x00007fff68d79458 abort_message + 231
4   libc++abi.dylib               	0x00007fff68d6a8a7 demangling_terminate_handler() + 238
5   libobjc.A.dylib               	0x00007fff6a8a75b1 _objc_terminate() + 104
6   libc++abi.dylib               	0x00007fff68d78887 std::__terminate(void (*)()) + 8
7   libc++abi.dylib               	0x00007fff68d78829 std::terminate() + 41
8   carta_backend                 	0x000000010adc8504 Session::~Session() + 404 (Session.cc:164)
9   carta_backend                 	0x000000010afd44ab SetCursorTask::~SetCursorTask() + 59 (OnMessageTask.h:45)
10  carta_backend                 	0x000000010afe5a6e void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, carta::ThreadManager::StartEventHandlingThreads(int)::$_0> >(void*) + 190 (thread:284)
11  libsystem_pthread.dylib       	0x00007fff6bc52109 _pthread_start + 148
12  libsystem_pthread.dylib       	0x00007fff6bc4db8b thread_start + 15

kswang1029 commented:
Here is another kind of crash when loading a cube twice and matching them and requesting region spectral profiles:

Segmentation fault: 11

Thread 3 Crashed:
0   libcasa_tables.5.dylib        	0x00000001098db956 casacore::TSMCube::accessLine(char*, unsigned int, unsigned int, bool, casacore::BucketCache*, casacore::IPosition const&, unsigned int, casacore::IPosition const&, unsigned int, unsigned int) + 2934
1   libcasa_tables.5.dylib        	0x00000001098dab14 casacore::TSMCube::accessSection(casacore::IPosition const&, casacore::IPosition const&, char*, unsigned int, unsigned int, unsigned int, bool) + 1972
2   libcasa_tables.5.dylib        	0x00000001098dbb9b casacore::TSMCube::accessStrided(casacore::IPosition const&, casacore::IPosition const&, casacore::IPosition const&, char*, unsigned int, unsigned int, unsigned int, bool) + 91
3   libcasa_tables.5.dylib        	0x00000001098eaaa1 casacore::TiledFileAccess::get(casacore::Array<float>&, casacore::Slicer const&) + 209
4   libcasa_images.5.dylib        	0x0000000109dc842a casacore::FITSImage::doGetSlice(casacore::Array<float>&, casacore::Slicer const&) + 234
5   libcasa_lattices.5.dylib      	0x000000010a076745 casacore::Lattice<float>::getSlice(casacore::Array<float>&, casacore::Slicer const&, bool) + 117
6   libcasa_lattices.5.dylib      	0x000000010a084d54 casacore::SubLattice<float>::doGetSlice(casacore::Array<float>&, casacore::Slicer const&) + 420
7   libcasa_lattices.5.dylib      	0x000000010a076745 casacore::Lattice<float>::getSlice(casacore::Array<float>&, casacore::Slicer const&, bool) + 117
8   libcasa_lattices.5.dylib      	0x000000010a059917 casacore::Lattice<float>::getSlice(casacore::Array<float>&, casacore::IPosition const&, casacore::IPosition const&, casacore::IPosition const&, bool) + 71
9   libcasa_lattices.5.dylib      	0x000000010a059607 casacore::LatticeIterInterface<float>::readData(bool) + 487
10  libcasa_lattices.5.dylib      	0x000000010a05911e casacore::LatticeIterInterface<float>::cursor(bool, bool) + 62
11  carta_backend                 	0x0000000106eca3b7 carta::FileLoader::GetSlice(casacore::Array<float>&, casacore::Slicer const&) + 375 (FileLoader.cc:322)
12  carta_backend                 	0x0000000106e8b17d Frame::GetSlicerData(casacore::Slicer const&, std::__1::vector<float, std::__1::allocator<float> >&) + 141
13  carta_backend                 	0x0000000106e935f3 Frame::FillSpectralProfileData(std::__1::function<void (CARTA::SpectralProfileData)>, int, bool) + 1763 (Frame.cc:1450)
14  carta_backend                 	0x0000000106e71f14 Session::SendSpectralProfileData(int, int, bool) + 516
15  carta_backend                 	0x0000000106e71515 Session::OnSetCursor(CARTA::SetCursor const&, unsigned int) + 1029
16  carta_backend                 	0x0000000107077bfe FileSettings::ExecuteOne(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int) + 382
17  carta_backend                 	0x00000001070762e8 SetCursorTask::execute() + 56 (OnMessageTask.cc:29)
18  carta_backend                 	0x0000000107087a65 void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, carta::ThreadManager::StartEventHandlingThreads(int)::$_0> >(void*) + 181 (thread:284)
19  libsystem_pthread.dylib       	0x00007fff6bc52109 _pthread_start + 148
20  libsystem_pthread.dylib       	0x00007fff6bc4db8b thread_start + 15

acdo2002 commented:
I have run the performance regression tests, no significant changes have been found. 
The performance regression tests are passed. 
The acdc server is back, the performance regression results can be found in the [link](https://carta.asiaa.sinica.edu.tw/performance/).

robsimmonds committed:
Changed to using shared pointers for the condition variables to
avoid possible race condition in their distruction.

robsimmonds committed:
Merge branch 'dev' into rwjs/cppmutex

robsimmonds committed:
Added missing counter decrement.

Went back to standard pointers for the CVs since I now don't think
they were causing a problem.

robsimmonds committed:
Merge branch 'dev' into rwjs/cppmutex

kswang1029 reviewed (approved):
I tested several use cases, including the two crash cases I encountered. No odd things are observed. I think this is ready to go. 👍
